runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -1092,24 +1107,39 @@ private void verifyAbsoluteTimeDomain() {
      * </ul>
      */
     private void setAndVerifyOutputTimestamp() {
-      // Output timestamp is currently not supported in processing time timers.
-      if (outputTimestamp != null && !TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
-        throw new IllegalStateException("Cannot set outputTimestamp in processing time domain.");
-      }
+
       // Output timestamp is set to the delivery time if not initialized by an user.
-      if (outputTimestamp == null) {
+      if (outputTimestamp == null && TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
         outputTimestamp = target;
       }
-
-      if (TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
-        Instant windowExpiry = window.maxTimestamp().plus(allowedLateness);
-        checkArgument(
-            !target.isAfter(windowExpiry),
-            "Attempted to set event time timer that outputs for %s but that is"
-                + " after the expiration of window %s",
-            target,
-            windowExpiry);
+      // For processing timers
+      if (outputTimestamp == null) {
+        // For processing timers output timestamp will be:
+        // 1) timestamp of input element
+        // OR
+        // 2) output timestamp of firing timer.
+        outputTimestamp = elementInputTimestamp;
       }
+
+      checkArgument(
+          !outputTimestamp.isBefore(elementInputTimestamp),
+          "output timestamp %s should be after input message timestamp or output timestamp of firing timers %s",
+          outputTimestamp,
+          elementInputTimestamp);
```

Comment:
check this before setting outputTimestamp above (and only if outputTimestamp != null)

Reply:
@reuvenlax  done